### 作业要求

 ![homework](homework.bmp)

In [2]:
from itertools import combinations, permutations

#### 求解系数的步骤：

In [3]:
multiply = lambda x,y:x*y
def coefficient(c,a):
    c_list = combinations(c,a)
    return (-1)**a*sum(map(lambda x:reduce(multiply,x),c_list))

In [4]:
print 'a0 到 a9的系数为:'
print list(reversed(map(coefficient,[range(1,11)]*10,range(1,11))))

a0 到 a9的系数为:
[3628800, -10628640, 12753576, -8409500, 3416930, -902055, 157773, -18150, 1320, -55]


由此知道返回的系数a0到a9的分别是以上的数字。则我们要解的方程是: 
$p(x) =  3628800-10628640*x+12753576*x^2-8409500*x^3+3416930*x^4-902055*x^5+157773*x^6+-18150*x^7+1320*x^8-55x+x^(10)$  
则扰动方程为:
$p(x)+ɛx^9=0$

 直接利用课本的(6.4)式子就可以求得解：
$x_k(\epsilon)\approx x_k-\frac{q(x_k)}{p'(x_k)}\epsilon$

知道$x_k=k(k=1,2,3\dots10)$  
$p'(x_k)= \prod_{j\neq k} (k-j)$

In [5]:
# 定义阶乘的函数
def factorial(k):
    if k==0:
        return 1
    else:
        return reduce(multiply,range(1,k+1))

In [6]:
factorial(8)

40320

In [7]:
def solve(k,e):
    zi = (-1)**(k-1)*e*(k**9)
    mu = factorial(k-1)*factorial(10-k)
    return k+zi*1.0/mu

In [8]:
e_list=[10**(-2),10**(-4),10**(-6)]

In [9]:
print map(solve,range(1,11),[0.01]*10)

[1.0000000275573193, 1.9998730158730158, 3.0195267857142856, 3.3931851851851853, 11.781684027777779, -28.992000000000004, 100.41112731481482, -125.15250793650793, 105.08643080357143, -17.55731922398589]


In [10]:
for e in e_list:
    print '当e为:',e
    print '10个解分别为:',
    print map(solve,range(1,11),[e]*10),'\n'

当e为: 0.01
10个解分别为: [1.0000000275573193, 1.9998730158730158, 3.0195267857142856, 3.3931851851851853, 11.781684027777779, -28.992000000000004, 100.41112731481482, -125.15250793650793, 105.08643080357143, -17.55731922398589] 

当e为: 0.0001
10个解分别为: [1.0000000002755731, 1.99999873015873, 3.000195267857143, 3.993931851851852, 5.067816840277778, 5.65008, 7.934111273148148, 6.668474920634921, 9.960864308035715, 9.72442680776014] 

当e为: 1e-06
10个解分别为: [1.0000000000027558, 1.9999999873015872, 3.0000019526785713, 3.9999393185185186, 5.000678168402778, 5.9965008, 7.009341112731481, 7.986684749206349, 9.009608643080357, 9.997244268077601] 



由此可得,这个$\epsilon$对整个方程的解的影响是非常的大的,尽管$\epsilon$到达10-6,影响也是无法忽略的.